ompi/mca/coll/sync/coll_sync.h

```diff
@@ -0,0 +1,183 @@
+                          struct ompi_communicator_t *comm,
+                          mca_coll_base_module_t *module);
+
+int mca_coll_sync_reduce(const void *sbuf, void *rbuf, int count,
+                         struct ompi_datatype_t *dtype,
+                         struct ompi_op_t *op,
+                         int root,
+                         struct ompi_communicator_t *comm,
+                         mca_coll_base_module_t *module);
+
+int mca_coll_sync_reduce_scatter(const void *sbuf, void *rbuf,
+                                 int *rcounts,
+                                 struct ompi_datatype_t *dtype,
+                                 struct ompi_op_t *op,
+                                 struct ompi_communicator_t *comm,
+                                 mca_coll_base_module_t *module);
+
+int mca_coll_sync_scan(const void *sbuf, void *rbuf, int count,
+                       struct ompi_datatype_t *dtype,
+                       struct ompi_op_t *op,
+                       struct ompi_communicator_t *comm,
+                       mca_coll_base_module_t *module);
+
+int mca_coll_sync_scatter(const void *sbuf, int scount,
+                          struct ompi_datatype_t *sdtype,
+                          void *rbuf, int rcount,
+                          struct ompi_datatype_t *rdtype,
+                          int root,
+                          struct ompi_communicator_t *comm,
+                          mca_coll_base_module_t *module);
+
+int mca_coll_sync_scatterv(const void *sbuf, int *scounts, int *disps,
+                           struct ompi_datatype_t *sdtype,
+                           void *rbuf, int rcount,
+                           struct ompi_datatype_t *rdtype,
+                           int root,
+                           struct ompi_communicator_t *comm,
+                           mca_coll_base_module_t *module);
+
+
+/* Types */
+/* Module */
+
+typedef struct mca_coll_sync_module_t {
+    mca_coll_base_module_t super;
+
+    /* Pointers to all the "real" collective functions */
+    mca_coll_base_comm_coll_t c_coll;
+
+    /* How many ops we've executed */
+    int before_num_operations;
+
+    /* How many ops we've executed (it's easier to have 2) */
+    int after_num_operations;
+
+    /* Avoid recursion of syncs */
+    bool in_operation;
+} mca_coll_sync_module_t;
+
+OBJ_CLASS_DECLARATION(mca_coll_sync_module_t);
+
+/* Component */
+
+typedef struct mca_coll_sync_component_t {
+    mca_coll_base_component_2_0_0_t super;
+
+    /* Priority of this component */
+    int priority;
+
+    /* Do a sync *before* each Nth collective */
+    int barrier_before_nops;
+
+    /* Do a sync *after* each Nth collective */
+    int barrier_after_nops;
+} mca_coll_sync_component_t;
+
+/* Globally exported variables */
+
+OMPI_MODULE_DECLSPEC extern mca_coll_sync_component_t mca_coll_sync_component;
+
+/* Macro used in most of the collectives */
+
+#define COLL_SYNC(module, op) \
```

Comment:
This is where we have an error - the ```module``` param should be used in the macro, but instead it uses an ```s``` variable. This works because all the functions created an ```s``` for the module, and so the macro picks that up. But it still isn't technically correct, and will fail if someone names the module something else in their code.